### ISSUE TITLE
aws_resource_missing_tags rule flags tagged resource aws_autoscaling_group

### ISSUE THREAD

**Issue by nitrocode** [terraform-linters/tflint#662]
```
TFLint v0.15.1
Terraform v0.12.23
```

In `~/.tflint.hcl`
```hcl
# missing tags
rule "aws_resource_missing_tags" {
  enabled = true
  tags    = ["Name", "env"]
}
```

terraform

```hcl
resource "aws_autoscaling_group" "this" {
  ...
  dynamic "tag" {
    for_each = local.tags

    content {
      key                 = tag.key
      value               = tag.value
      propagate_at_launch = true
    }
  }
}
```

Since `tags` is an argument and `tag` is an argument for resource [`aws_autoscaling_group`](https://www.terraform.io/docs/providers/aws/r/autoscaling_group.html), there needs to be additional logic for this resource to check if it has the correct values.

Or perhaps we can exclude resources from rules. Is there an exclusion argument for rules?

@bwhaley

**Comment by bwhaley** [comment]
Ahh, I didn't realize that `aws_autoscaling_group` approached tags differently. 

There are a couple of distinct issues here:

1. The rule should handle both the `tag` block and the `tags` attribute for `aws_autoscaling_group`. I have some prototype logic that seems to work, and once I finish that up I'll submit a pull request.
2. AFAICT, [`dynamic` blocks with the `for_each` iterator are not expanded](https://github.com/terraform-linters/tflint/blob/master/tflint/runner.go#L541), so those will be unevaluable. I might be missing something here. Perhaps @wata727 can comment on that.

**Comment by wata727** [comment]
Yes, dynamic blocks cannot be expanded, so they need to be treated specially. There are several options.

1. Ignore resources that have a dynamic `tag`. This is a design that allows false negatives and focuses on not reporting any extra issues.
2. If a resource has a dynamic `tag` but cannot verify that the key exists, emit an issue. This is a design that allows false positives and focuses on not missing out on problems.

I'm not sure which one is better, but probably the former is more acceptable to most users.

**Comment by nitrocode** [comment]
That's too bad about the `dynamic` block unable to be expanded. Is there a ticket for terraform to fix this limitation?

Would you be open to a `exclude` clause for particular resources? or for now, should we just remove the `aws_autoscaling_group` from the list of resources that the rule checks for until we have better logic for the `tags` argument?

https://github.com/terraform-linters/tflint/blob/master/rules/awsrules/aws_resource_missing_tags.go#L48

**Comment by wata727** [comment]
> Is there a ticket for terraform to fix this limitation?

No. In fact, it is technically possible to expand dynamic blocks. However, if we use the same way as Terraform, all values in the dynamic block must be *known*.

TFLint has a mechanism that can inspect partially known values while keeping many unknown values, such as the values referenced from tfstate. However, because dynamic blocks expansion evaluates the entire block, I believe that a mechanism to evaluate only a part of the dynamic block is necessary to perform a practical inspection.

Probably possible, but pending many considerations.
